module3/lessons/testing_tools_for_api_consumption.md

```diff
@@ -0,0 +1,468 @@
+
+      expect(json[:data][:id]).to be_nil
+      expect(json[:data][:type]).to eq("image")
+      expect(json[:data][:attributes]).to have_key(:image_url)
+      expect(json[:data][:attributes]).to have_key(:photographer)
+      expect(json[:data][:attributes]).to have_key(:photographer_url)
+      expect(json[:data][:attributes]).to have_key(:alt_text)
+    end
+  end
+end
+```
+
+We can put this right above the `get "/api/v1/images?artist=The%20Beatles"`.
+
+Now, when we run the test, we get a new error.
+
+
+```bash
+Failures:
+
+  1) Images Endpoint happy path can retrieve an image for a specific artist specific artist
+     Failure/Error: json = JSON.parse(response.body, symbolize_names: true)
+     
+     JSON::ParserError:
+       unexpected token at ''
+     # ./app/controllers/api/v1/images_controller.rb:11:in `show'
+     # ./spec/requests/api/v1/images/image_request_spec.rb:9:in `block (3 levels) in <top (required)>'
+```
+
+If we look at the stub we just put in the test, we are returning an empty body, so it makes sense that we're getting an error when trying to parse the response body as JSON.
+
+We need to replace the empty body with an actual JSON response. We *could* copy and paste a body right into this test, but then our test file would get quite messy. What we'll do instead is make a `spec/fixtures` directory with a file that we can read:
+
+```bash
+$ mkdir spec/fixtures
+$ touch spec/fixtures/beatles_artist_query.json
+```
+
+And then we have to update our test to use this fixture file:
+
+*spec/requests/api/v1/images/image_request_spec.rb*
+
+```ruby
+require "rails_helper"
+
+RSpec.describe "Images Endpoint" do
+  describe "happy path" do
+    it "can retrieve an image for a specific artist specific artist" do
+      json_response = File.read('spec/fixtures/beatles_artist_query.json')
+      stub_request(:get, "https://api.pexels.com/v1/search?query=The%20Beatles")
+        .to_return(status: 200, body: json_response)
+
+      get "/api/v1/images?artist=The%20Beatles"
+
+      expect(response).to be_successful
+      json = JSON.parse(response.body, symbolize_names: true)
+
+      expect(json[:data][:id]).to be_nil
+      expect(json[:data][:type]).to eq("image")
+      expect(json[:data][:attributes]).to have_key(:image_url)
+      expect(json[:data][:attributes]).to have_key(:photographer)
+      expect(json[:data][:attributes]).to have_key(:photographer_url)
+      expect(json[:data][:attributes]).to have_key(:alt_text)
+    end
+  end
+end
+```
+
+We're still returning an empty body because our file is empty, so let's add some actual JSON data to that file **That mimics how the JSON Data looks when we hit the real API**. Use Postman to hit the Pexels API to get a JSON response and copy and paste it into your *spec/fixtures/beatles_artist_query.json* file. Your test should be passing once again.
+
+If this is *really* working, we should be able to turn off our WiFi and see the test is still working.
+
```

Comment:
I think I would do a pro/con discussion questions here before we get into VCR.  I think it would be good for them to 'discover' how using external apis (with stubs or otherwise) forces us to rely on the stability of other's applications.